sdk/scheduler/src/main/java/com/mesosphere/sdk/curator/CuratorPersister.java

```diff
@@ -266,30 +254,42 @@ public void setMany(Map<String, byte[]> unprefixedPathBytesMap) throws Persister
+        // Convert map to translated prefixed paths. Manually construct TreeMap for consistent ordering (for tests):
+        Map<String, byte[]> pathBytesMap = new TreeMap<>();
         for (Map.Entry<String, byte[]> entry : unprefixedPathBytesMap.entrySet()) {
             pathBytesMap.put(withFrameworkPrefix(entry.getKey()), entry.getValue());
         }
-        logger.debug("Setting many entries: {}", pathBytesMap.keySet());
+        logger.debug("Updating many entries: {}", pathBytesMap.keySet());
         try {
             for (int i = 0; i < ATOMIC_WRITE_ATTEMPTS; ++i) {
-                // Phase 1: Determine which nodes already exist. This determination can be rendered
-                //          invalid by an out-of-band change to the data.
-                //TODO(nickbp): This currently doesn't correctly detect when our own transaction is creating a node.
-                //              For example, a transaction which writes both "/a" and "/a/b" entries will not detect
-                //              that "/a" will already exist by the time "/a/b" is being written. However in practice
-                //              this hasn't come up (yet...)
-                final Set<String> pathsWhichExist = selectPathsWhichExist(pathBytesMap.keySet());
-                List<String> parentPathsToCreate = getParentPathsToCreate(pathBytesMap.keySet(), pathsWhichExist);
-
-                logger.debug("Atomic write attempt {}/{}:\n-Parent paths: {}\n-All paths: {}\n-Paths which exist: {}",
-                        i + 1, ATOMIC_WRITE_ATTEMPTS, parentPathsToCreate, pathBytesMap.keySet(), pathsWhichExist);
-
-                // Phase 2: Compose a single transaction that updates and/or creates nodes according to
-                //          the above determinations (retry if there's an out-of-band modification)
-                final CuratorTransactionFinal transaction =
-                        getWriteTransaction(pathBytesMap, pathsWhichExist, parentPathsToCreate);
+                // List of paths that are known to exist, or which are about to be created by the transaction
+                // Includes "known to exist" in order to avoid repeated lookups for the same path
+                Set<String> existingAndPendingCreatePaths = new HashSet<>();
+                // List of paths which are about to be deleted by the transaction
+                Set<String> pendingDeletePaths = new HashSet<>();
+
+                CuratorTransactionFinal transaction = client.inTransaction().check().forPath(serviceRootPath).and();
+                for (Map.Entry<String, byte[]> entry : pathBytesMap.entrySet()) {
+                    String path = entry.getKey();
+                    if (entry.getValue() == null) {
+                        // null value: if present, delete path and any children (unless already being deleted)
+                        if (!pendingDeletePaths.contains(path)
```

Comment:
The logic here gets pretty hairy at this level of nesting, and considering that it concerns something as fundamental as ZK, I think it'd be worth breaking apart for readability.

Reply:
Good point, broke out the transaction creation itself into a separate function. Much cleaner now.